[mlir][vector] Update representation of insert/extract_strided_slice

MacDue opened the pull request:
This commit updates the representation of both extract_strided_slice and insert_strided_slice to primitive arrays of int64_ts, rather than ArrayAttrs of IntegerAttrs. This prevents a lot of boilerplate conversions between IntegerAttr and int64_t.

This is done by adding a new `StridedSliceAttr` which matches the previous syntax and can be used for both operations.

It may also be possible to explore alternate slice syntax for the `StridedSliceAttr` in future.

MacDue commented:
This is the most ambitious change based on #100997. 

**Note:** This PR is not urgent! It's just something I think would be nice to have. :slightly_smiling_face:

joker-eph commented on mlir/include/mlir/Dialect/Vector/IR/VectorAttributes.td:
```diff
@@ -82,4 +87,63 @@ def Vector_PrintPunctuation : EnumAttr<Vector_Dialect, PrintPunctuation, "punctu
+    - The first dimension starts at offset 0 and is strided by 1
+    - The second dimension starts at offset 4 and is strided by 2
+
+    `[0, 1, 2][3:1][4:8]`
```
I find this quite hard to read: why are some dimensions separated in their own `[` `]` group but not the non-strided ones?

In general the implicit size seems also hard to read. I'm not sure how to make it better though right now.

Looking at some examples that go from `{offsets = [1, 1], sizes = [2, 2], strides = [1, 1]}` to `[1:2:1][1:2:1]`, I appreciate the compact aspect but I'm not looking forward to need to keep the doc open always to be able to read it...

MacDue replied:
I admit it's not the most readable (I was trying to match the TODO first). 

I grouped the non-strided offsets so they'd look like a normal extraction vs the strided dims (though from updating the tests multiple non-strided dims do not appear to occur in practice much -- the only case IIRC is in the invalid tests).

---

Possibly, a simpler format could be something like:

`[0, 1, 2, 3:1, 4:8]`

And with a size:

`[1:1, 1:1] size 2x2`

i.e. a list of offsets, optionally followed by a `: stride`, with the size displayed separately.

So,
```mlir
%1 = vector.extract_strided_slice %0[0:2:1][2:4:1] : vector<4x8x16xf32> to vector<2x4x16xf32>
```
Becomes:
```mlir
%1 = vector.extract_strided_slice %0[0:1, 2:1] size 2x4 : vector<4x8x16xf32> to vector<2x4x16xf32>
```
---
And,
```mlir
%2 = vector.insert_strided_slice %0, %1[0][0:1][2:1] : vector<2x4xf32> into vector<16x4x8xf32>
```

Becomes:
```mlir
%2 = vector.insert_strided_slice %0, %1[0, 0:1, 2:1] : vector<2x4xf32> into vector<16x4x8xf32>
```

banach-space replied:
> In general the implicit size seems also hard to read. I'm not sure how to make it better though right now.

I don't have any good suggestion myself. Personally, I find the existing syntax clearer.

hanhanW replied:
I think the existing syntax is clearer too. My first thought about `[1:2:1]` is something like 
"iterate on `range(1, 2)` with `strides=1`", though it does not make sense to insert/extract_strided_slice ops. Spelling out `offsets`, `sizes`, and `strides` looks clearer to me.

On the other hand, the existing style looks more consistent with the tensor version. E.g., `tensor.insert_slice %11 into %arg4[%8, 0] [1, 2] [1, 1]`

MacDue replied:
I realised I could preserve the old syntax pretty much exactly with a custom `assemblyFormat`. Updated the patch now.

joker-eph commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -2899,6 +2883,95 @@ OpFoldResult vector::InsertOp::fold(FoldAdaptor adaptor) {
+      printer << sizes[strideIdx] << ':';
+    printer << strides[strideIdx] << ']';
+  }
+}
```
There seems to be an assumption about the size of all the arrays, but I don't see the invariant enforced in a verifier?

MacDue replied:
I didn't add a verifier for the attribute, as the constraints are enforced within the ops (right now). I think some of those constraints could be factored out to the `StridedSliceAttr` though :+1:

MacDue replied:
Going to leave this for a future patch now it's not needed for the printer/parser (which just uses the current syntax).

MacDue committed:
[mlir][vector] Update representation of insert/extract_strided_slice

This commit updates the representation of both extract_strided_slice and
insert_strided_slice to primitive arrays of int64_ts, rather than
ArrayAttrs of IntegerAttrs. This prevents a lot of boilerplate
conversions between IntegerAttr and int64_t.

This is done by adding a new `StridedSliceAttr` which matches the
previous syntax and can be used for both operations.

It may also be possible to explore alternate slice syntax for the
`StridedSliceAttr` in future.
